cpp/src/unary/cast_ops.cu

```diff
@@ -25,47 +25,104 @@
 
 namespace cudf {
 namespace detail {
-template <typename _T, typename _R>
+template <typename _TargetT>
 struct unary_cast {
-  template <typename T                                                                   = _T,
-            typename R                                                                   = _R,
-            typename std::enable_if_t<(cudf::is_numeric<T>() && cudf::is_numeric<R>())>* = nullptr>
-  CUDA_DEVICE_CALLABLE R operator()(T const element)
+  template <typename SourceT,
+            typename TargetT                                          = _TargetT,
+            typename std::enable_if_t<(cudf::is_numeric<SourceT>() &&
+                                       cudf::is_numeric<TargetT>())>* = nullptr>
+  CUDA_DEVICE_CALLABLE TargetT operator()(SourceT const element)
   {
-    return static_cast<R>(element);
+    return static_cast<TargetT>(element);
   }
-  template <
-    typename T                                                                       = _T,
-    typename R                                                                       = _R,
-    typename std::enable_if_t<(cudf::is_timestamp<T>() && cudf::is_timestamp<R>())>* = nullptr>
-  CUDA_DEVICE_CALLABLE R operator()(T const element)
+
+  template <typename SourceT,
+            typename TargetT                                            = _TargetT,
+            typename std::enable_if_t<(cudf::is_timestamp<SourceT>() &&
+                                       cudf::is_timestamp<TargetT>())>* = nullptr>
+  CUDA_DEVICE_CALLABLE TargetT operator()(SourceT const element)
+  {
+    // Convert source tick counts into target tick counts without blindly truncating them
+    // by dividing the respective duration time periods (which may not work for time before
+    // UNIX epoch)
+    return TargetT{simt::std::chrono::floor<TargetT::duration>(element.time_since_epoch())};
+  }
+
+  template <typename SourceT,
+            typename TargetT                                           = _TargetT,
+            typename std::enable_if_t<(cudf::is_duration<SourceT>() &&
+                                       cudf::is_duration<TargetT>())>* = nullptr>
+  CUDA_DEVICE_CALLABLE TargetT operator()(SourceT const element)
+  {
+    return TargetT{simt::std::chrono::floor<TargetT>(element)};
+  }
+
+  // TODO: Should this be allowed? timestamps should be constructed from duration or another
+  // timestamp. Will this break existing code that rely on this functionality?
```

Comment:
while i was working on supporting the `datediff` functionality earlier, there were objections to allow casting timestamps to integers (basically, a binary operator that subtracts 2 timestamps and returns an integer). i see here that we do allow it via the casting operator, and thus the question. if there are users that rely on this behavior already, then it may be difficult to change (or disallow it).

Reply:
Subtracting two timestamp column in Python should return a duration column, and then (I assume) the duration could be cast to the underlying numeric type if needed.

Reply:
yes, the former will be supported as a part of this work. however, what do we do with these (conversions between timestamp and numeric)? this could be a breaking change for consumers that use this already.

Reply:
Breaking changes are fine, we do them all the time.